Refactor initGenesis function

aidan-kwon opened the pull request:
## Proposed changes

`initGenesis` does followings:
 - load and validate genesis config
 - set governance data with loaded data
 - prepare database 
 - store genesis state in database

This PR refined `initGenesis` to be a more readable code. 
`checkGenesisAndFillDefaultIfNeeded` and other codes are split into validation part and initiation part (setDefault). 
Some validation logics covering conner cases are introduced also. 

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

ehnuje commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -232,45 +212,53 @@ func writeFailLog(key int, err error) {
 	}
 
-	// using Clique as a consensus engine
-	if genesis.Config.Istanbul == nil && genesis.Config.Clique != nil {
```
Is this line unnecessary in the refactored version? I couldn't find this line.

aidan-kwon replied:
These lines match up with line 269 ~ 273 in the refactored version.

ehnuje replied:
```
if c.Governance == nil {
	engineType := c.GetConsensusEngine()
	c.Governance = GetDefaultGovernanceConfig(engineType)
	logger.Warn("Override the default governance config to the chain config", "engineType", engineType)
}
```
Do you mean above codes?

aidan-kwon replied:
Yes. This `if` branch and the next `else if` branch set default governance config if it doesn't exist. The above code you quoted does the same behaviors.

ehnuje commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -232,45 +212,53 @@ func writeFailLog(key int, err error) {
 
-	if valueChanged {
-		logger.Warn("Some input value of genesis.json have been set to default or changed")
+	if g.Config.Governance == nil || g.Config.Governance.Reward == nil {
```
Is this a newly added line?

aidan-kwon replied:
Yes. I forgot to notice that some validations are introduced.

yoomee1313 replied:
The below error tells that governance and reward policy should be configured. However, if-loop contains or (||)!

jeongkyun-oh commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -232,45 +212,53 @@ func writeFailLog(key int, err error) {
-func checkGenesisAndFillDefaultIfNeeded(genesis *blockchain.Genesis) *blockchain.Genesis {
-	engine := params.UseIstanbul
-	valueChanged := false
-	if genesis.Config == nil {
```
Is removing this nil check on purpose?

jeongkyun-oh commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -74,78 +75,57 @@ func initGenesis(ctx *cli.Context) error {
-			}
-		}
+	if genesis.Config == nil {
+		logger.Crit("Genesis config is not set")
```
Shouldn't it set default configuration accroding to line # 239?

aidan-kwon replied:
The original code allowed creating a new `genesis.Config`. However, it was not a safe way since the new `genesis.Config` requires a `ChainID` as mandatory. So, I invalidate that case.

jeongkyun-oh commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -232,45 +212,53 @@ func writeFailLog(key int, err error) {
-		// Error case. Both istanbul and Clique exists
-		logger.Crit("Both clique and istanbul configuration exists. Only one configuration can be applied. Exiting..")
+	if g.Config.Clique == nil && g.Config.Istanbul == nil {
+		return errors.New("consensus engine should be configured")
```
It seems that the default engine is `UseIstanbul` if no engine is set. Do you have any reason to remove the default setting?

ehnuje commented:
> initGenesis does followings
> 
> - load and validate genesis config
> - set governance data with loaded data
> - prepare database
> - store genesis state in database

@aidan-kwon Above PR description does not explain things done in this PR. Could you please elaborate the detail of this PR in the description?

aidan-kwon commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -232,45 +212,53 @@ func writeFailLog(key int, err error) {
 	logger.Crit(msg, "err", err)
 }
 
-func checkGenesisAndFillDefaultIfNeeded(genesis *blockchain.Genesis) *blockchain.Genesis {
```
`checkGenesisAndFillDefaultIfNeeded` is split into `genesis.Config.SetDefaults()` and `ValidateGenesisConfig(genesis)`.

aidan-kwon commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -74,78 +75,57 @@ func initGenesis(ctx *cli.Context) error {
-
-	genesis = checkGenesisAndFillDefaultIfNeeded(genesis)
-
-	if genesis.Config.Istanbul != nil {
-		if err := governance.CheckGenesisValues(genesis.Config); err != nil {
-			logger.Crit("Error in genesis json values", "err", err)
-		}
-
-		// Check if governingNode is properly set
-		if strings.ToLower(genesis.Config.Governance.GovernanceMode) == "single" {
-			if istanbulExtra, err := decodeExtra(genesis.ExtraData); err != nil {
-				logger.Crit("Extra data couldn't be decoded. Please check if your extra data is correct", "err", err)
-			} else {
-				var found bool
-				for _, v := range istanbulExtra.Validators {
-					if v == genesis.Config.Governance.GoverningNode {
-						found = true
-						break
-					}
-				}
-				if !found {
-					logger.Crit("GoverningNode is not in the validator list. Please check if your governingNode address is correct")
-				}
-			}
-		}
```
Thes lines are moved to the inside of `ValidateGenesisConfig`

aidan-kwon commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -74,78 +75,57 @@ func initGenesis(ctx *cli.Context) error {
 
-	if genesis.Config.Governance.Reward.StakingUpdateInterval != 0 {
-		params.SetStakingUpdateInterval(genesis.Config.Governance.Reward.StakingUpdateInterval)
-	} else {
-		genesis.Config.Governance.Reward.StakingUpdateInterval = params.StakingUpdateInterval()
-	}
+	// Update undefined config with default values
+	genesis.Config.SetDefaults()
 
-	if genesis.Config.Governance.Reward.ProposerUpdateInterval != 0 {
-		params.SetProposerUpdateInterval(genesis.Config.Governance.Reward.ProposerUpdateInterval)
-	} else {
-		genesis.Config.Governance.Reward.ProposerUpdateInterval = params.ProposerUpdateInterval()
```
These line are split into `genesis.Config.SetDefaults()` and `ValidateGenesisConfig()`.

aidan-kwon commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -74,78 +75,57 @@ func initGenesis(ctx *cli.Context) error {
-	data := getGovernanceItemsFromGenesis(genesis)
-	gbytes, err := json.Marshal(data.Items())
+	// Set genesis.Governance and reward intervals
+	govSet := getGovernanceItemSetFromGenesis(genesis)
+	govItemBytes, err := json.Marshal(govSet.Items())
 	if err != nil {
 		logger.Crit("Failed to json marshaling governance data", "err", err)
 	}
-	if genesis.Governance, err = rlp.EncodeToBytes(gbytes); err != nil {
```
Just renamed some variables

aidan-kwon commented on params/config.go:
```diff
@@ -242,6 +244,53 @@ func (c *ChainConfig) checkCompatible(newcfg *ChainConfig, head *big.Int) *Confi
 
+// GetConsensusEngine returns the consensus engine type specified in ChainConfig.
+// It returns Unknown type if none of engine type is configured or more than one type is configured.
+func (c *ChainConfig) GetConsensusEngine() EngineType {
```
The newly introduced method to get consensus engine type

aidan-kwon commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -232,45 +212,53 @@ func writeFailLog(key int, err error) {
-	if genesis.Config == nil {
-		genesis.Config = new(params.ChainConfig)
+func ValidateGenesisConfig(g *blockchain.Genesis) error {
+	if g.Config.ChainID == nil {
+		return errors.New("chainID is not specified")
```
Newly introduced validation which was not covered before

aidan-kwon commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -232,45 +212,53 @@ func writeFailLog(key int, err error) {
 
-	if valueChanged {
-		logger.Warn("Some input value of genesis.json have been set to default or changed")
+	if g.Config.Governance == nil || g.Config.Governance.Reward == nil {
+		return errors.New("governance and reward policies should be configured")
```
Newly introduced validation which was not covered before

aidan-kwon commented on cmd/utils/nodecmd/chaincmd.go:
```diff
@@ -232,45 +212,53 @@ func writeFailLog(key int, err error) {
-	istanbulExtra, err := types.ExtractIstanbulExtra(&types.Header{Extra: extraData})
-	if err != nil {
-		return nil, err
+	if g.Config.Governance.Reward.ProposerUpdateInterval == 0 || g.Config.Governance.Reward.
+		StakingUpdateInterval == 0 {
+		return errors.New("proposerUpdateInterval and stakingUpdateInterval cannot be zero")
+	}
```
Newly introduced validation which was not covered before

aidan-kwon commented:
> @aidan-kwon Above PR description does not explain things done in this PR. Could you please elaborate the detail of this PR in the description?

I updated the PR description and added comments for more detailed explanations.

aidan-kwon commented:
@winnie-byun @KimKyungup PTAL

winnie-byun reviewed (approved):
LGTM :)
